Automattic/wp-calypso
# Checkout: Move analytics for redirect payment methods into payment processors

#### Changes proposed in this Pull Request

Because of the way in which composite checkout payment methods have evolved, all of the work, including recording analytics, has historically been inside the payment method object's submit button. Since payment method objects were originally defined inside a package outside of calypso, the `useEvents` hook was added to the package as a means to arbitrarily record analytics.

More recently, the responsibility for processing a transaction has moved to the payment processor functions. Since the payment processor functions already live inside calypso, they don't need the leaky abstraction of `useEvents` and can record analytics directly. This PR alters the redirect payment methods to move the analytics into the processor functions. Ultimately this will make the processor functions (and composite checkout as a whole) more flexible.

#### Testing instructions

To test, we should first make sure that each payment processor still works, but also that its analytics are recorded. To do this run `localStorage.setItem('debug', 'calypso:analytics')` in your browser console and reload the page before completing checkout. Verify that you see events for `calypso_checkout_form_redirect`, `calypso_checkout_form_submit`, `calypso_checkout_composite_form_submit`, and `calypso_checkout_composite_X_submit_clicked` (where `X` is the payment method id).

Since all these payment methods use the same payment processor function, they will all need to be tested in this one PR.

- WeChat: see https://github.com/Automattic/wp-calypso/pull/43765
- NetBanking: see https://github.com/Automattic/wp-calypso/pull/44952
- IdWallet: see https://github.com/Automattic/wp-calypso/pull/45257
- EbanxTef: see https://github.com/Automattic/wp-calypso/pull/44752
- Sofort: https://github.com/Automattic/wp-calypso/pull/43710
- P24: https://github.com/Automattic/wp-calypso/pull/43744
- Ideal: https://github.com/Automattic/wp-calypso/pull/43162
- Giropay: https://github.com/Automattic/wp-calypso/pull/43598
- Eps: see https://github.com/Automattic/wp-calypso/pull/43734
- BanContact: see https://github.com/Automattic/wp-calypso/pull/43651
- Alipay: see https://github.com/Automattic/wp-calypso/pull/43653
- PayPal: https://github.com/Automattic/wp-calypso/pull/46431

We also need to verify that the `calypso_checkout_composite_error` fires if there is an error in the analytics code. To test this, add something like `throw new Error('testing');` into `recordTransactionBeginAnalytics` inside the `try` block. Then try to complete any redirect payment method transaction and verify that you see `calypso_checkout_composite_error` fired.
